File: Leetcode-solutions/Binary-search/MedianOfTwoSortedArrays.java
public class MedianOfTwoSortedArrays {
    public static void main(String[] args) {

    }

    static double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if(nums2.length < nums1.length){
            return findMedianSortedArrays(nums2, nums1);
        }

        int s = 0;
        int e = nums1.length;

        while(s <= e){
            int cut1 = s + (e-s)/2;
            int cut2 = (nums1.length + nums2.length) /2 - cut1;

            int l1 = cut1 <= 0 ? Integer.MIN_VALUE : nums1[cut1 - 1];
            int l2 = cut2 <= 0 ? Integer.MIN_VALUE : nums2[cut2 - 1];

            int r1 = cut1 >= nums1.length ? Integer.MAX_VALUE : nums1[cut1];
            int r2 = cut2 >= nums2.length ? Integer.MAX_VALUE : nums2[cut2];

            if( l1 > r2 ){
                e = cut1 - 1;
            } else if( l2 > r1 ){
                s = cut1 + 1;
            } else{
                return (nums1.length + nums2.length) % 2 == 0 ?
                        (float) ( Math.max(l1, l2) + Math.min(r1, r2) ) / 2 :
                        Math.min(r1, r2);
            }

        }
        return -1;
    }
}

File: CPP/Problems/Median of Two Sorted Arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        vector<int> nums;
        int n1 = nums1.size();
        int n2 = nums2.size();
        int i = 0, j = 0;
        while(i<n1 and j<n2){
            if(nums1[i] <= nums2[j]){
                nums.push_back(nums1[i++]);
            }
            else{
                nums.push_back(nums2[j++]);
            }
        }
        cout<<i<<j;
        while(i<n1){
            nums.push_back(nums1[i++]);
        }
        while(j<n2){
            nums.push_back(nums2[j++]);
        }
        cout<<i<<j;
        int n = nums.size()/2;
        cout<<nums.size();
        if(nums.size()%2 == 0){
            cout<<(nums[n-1]+nums[n]);
            return (double)(nums[n-1]+nums[n])/2;
        }
        else return (double)(nums[n]);
    }
};
